# 직접 해보세요!
## melt 메서드 사용하기(151쪽)

### 1. 1개의 열만 고정하고 나머지 열을 행으로 바꾸기

In [1]:
import pandas as pd
pew = pd.read_csv('../data/pew.csv')
pew.head()

religion  <$10k  $10-20k  $20-30k  $30-40k  $40-50k  $50-75k  \
0            Agnostic     27       34       60       81       76      137   
1             Atheist     12       27       37       52       35       70   
2            Buddhist     27       21       30       34       33       58   
3            Catholic    418      617      732      670      638     1116   
4  Don’t know/refused     15       14       15       11       10       35   

   $75-100k  $100-150k  >150k  Don't know/refused  
0       122        109     84                  96  
1        73         59     74                  76  
2        62         39     53                  54  
3       949        792    633                1489  
4        21         17     18                 116

In [2]:
print(pew.iloc[:, 0:6])

                   religion  <$10k  $10-20k  $20-30k  $30-40k  $40-50k
0                  Agnostic     27       34       60       81       76
1                   Atheist     12       27       37       52       35
2                  Buddhist     27       21       30       34       33
3                  Catholic    418      617      732      670      638
4        Don’t know/refused     15       14       15       11       10
5          Evangelical Prot    575      869     1064      982      881
6                     Hindu      1        9        7        9       11
7   Historically Black Prot    228      244      236      238      197
8         Jehovah's Witness     20       27       24       24       21
9                    Jewish     19       19       25       25       30
10            Mainline Prot    289      495      619      655      651
11                   Mormon     29       40       48       51       56
12                   Muslim      6        7        9       10        9
13    

In [3]:
pew_long = pd.melt(pew, id_vars='religion')
pew_long

religion            variable  value
0                   Agnostic               <$10k     27
1                    Atheist               <$10k     12
2                   Buddhist               <$10k     27
3                   Catholic               <$10k    418
4         Don’t know/refused               <$10k     15
5           Evangelical Prot               <$10k    575
6                      Hindu               <$10k      1
7    Historically Black Prot               <$10k    228
8          Jehovah's Witness               <$10k     20
9                     Jewish               <$10k     19
10             Mainline Prot               <$10k    289
11                    Mormon               <$10k     29
12                    Muslim               <$10k      6
13                  Orthodox               <$10k     13
14           Other Christian               <$10k      9
15              Other Faiths               <$10k     20
16     Other World Religions               <$10k      5
17              Unaffiliated               <$10k    217
18                  Agnostic             $10-20k     34
19                   Atheist             $10-20k     27
20                  Buddhist             $10-20k     21
21                  Catholic             $10-20k    617
22        Don’t know/refused             $10-20k     14
23          Evangelical Prot             $10-20k    869
24                     Hindu             $10-20k      9
25   Historically Black Prot             $10-20k    244
26         Jehovah's Witness             $10-20k     27
27                    Jewish             $10-20k     19
28             Mainline Prot             $10-20k    495
29                    Mormon             $10-20k     40
..                       ...                 ...    ...
150                    Hindu               >150k     54
151  Historically Black Prot               >150k     78
152        Jehovah's Witness               >150k      6
153                   Jewish               >150k    151
154            Mainline Prot               >150k    634
155                   Mormon               >150k     42
156                   Muslim               >150k      6
157                 Orthodox               >150k     46
158          Other Christian               >150k     12
159             Other Faiths               >150k     41
160    Other World Religions               >150k      4
161             Unaffiliated               >150k    258
162                 Agnostic  Don't know/refused     96
163                  Atheist  Don't know/refused     76
164                 Buddhist  Don't know/refused     54
165                 Catholic  Don't know/refused   1489
166       Don’t know/refused  Don't know/refused    116
167         Evangelical Prot  Don't know/refused   1529
168                    Hindu  Don't know/refused     37
169  Historically Black Prot  Don't know/refused    339
170        Jehovah's Witness  Don't know/refused     37
171                   Jewish  Don't know/refused    162
172            Mainline Prot  Don't know/refused   1328
173                   Mormon  Don't know/refused     69
174                   Muslim  Don't know/refused     22
175                 Orthodox  Don't know/refused     73
176          Other Christian  Don't know/refused     18
177             Other Faiths  Don't know/refused     71
178    Other World Religions  Don't know/refused      8
179             Unaffiliated  Don't know/refused    597

[180 rows x 3 columns]

In [4]:
pew_long = pd.melt(pew, id_vars='religion', var_name='income', value_name='count')
print(pew_long.head())

             religion income  count
0            Agnostic  <$10k     27
1             Atheist  <$10k     12
2            Buddhist  <$10k     27
3            Catholic  <$10k    418
4  Don’t know/refused  <$10k     15


### 5. 2개 이상의 열을 고정하고 나머지 열을 행으로 바꾸기

In [5]:
billboard = pd.read_csv('../data/billboard.csv')

billboard.iloc[0:5, 0:16]

year        artist                    track  time date.entered  wk1   wk2  \
0  2000         2 Pac  Baby Don't Cry (Keep...  4:22   2000-02-26   87  82.0   
1  2000       2Ge+her  The Hardest Part Of ...  3:15   2000-09-02   91  87.0   
2  2000  3 Doors Down               Kryptonite  3:53   2000-04-08   81  70.0   
3  2000  3 Doors Down                    Loser  4:24   2000-10-21   76  76.0   
4  2000      504 Boyz            Wobble Wobble  3:35   2000-04-15   57  34.0   

    wk3   wk4   wk5   wk6   wk7   wk8   wk9  wk10  wk11  
0  72.0  77.0  87.0  94.0  99.0   NaN   NaN   NaN   NaN  
1  92.0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
2  68.0  67.0  66.0  57.0  54.0  53.0  51.0  51.0  51.0  
3  72.0  69.0  67.0  65.0  55.0  59.0  62.0  61.0  61.0  
4  25.0  17.0  17.0  31.0  36.0  49.0  53.0  57.0  64.0

In [6]:
billboard_long = pd.melt(billboard, id_vars=['year', 'artist', 'track', 'time', 'date.entered'], var_name='week', value_name='rating')

billboard_long

year                artist                    track  time date.entered  \
0      2000                 2 Pac  Baby Don't Cry (Keep...  4:22   2000-02-26   
1      2000               2Ge+her  The Hardest Part Of ...  3:15   2000-09-02   
2      2000          3 Doors Down               Kryptonite  3:53   2000-04-08   
3      2000          3 Doors Down                    Loser  4:24   2000-10-21   
4      2000              504 Boyz            Wobble Wobble  3:35   2000-04-15   
5      2000                  98^0  Give Me Just One Nig...  3:24   2000-08-19   
6      2000               A*Teens            Dancing Queen  3:44   2000-07-08   
7      2000               Aaliyah            I Don't Wanna  4:15   2000-01-29   
8      2000               Aaliyah                Try Again  4:03   2000-03-18   
9      2000        Adams, Yolanda            Open My Heart  5:30   2000-08-26   
10     2000         Adkins, Trace                     More  3:05   2000-04-29   
11     2000   Aguilera, Christina  Come On Over Baby (A...  3:38   2000-08-05   
12     2000   Aguilera, Christina            I Turn To You  4:00   2000-04-15   
13     2000   Aguilera, Christina        What A Girl Wants  3:18   1999-11-27   
14     2000          Alice Deejay         Better Off Alone  6:50   2000-04-08   
15     2000           Allan, Gary  Smoke Rings In The D...  4:18   2000-01-22   
16     2000                 Amber                   Sexual  4:38   1999-07-17   
17     2000             Anastacia           I'm Outta Love  4:01   2000-04-01   
18     2000         Anthony, Marc              My Baby You  3:59   2000-09-16   
19     2000         Anthony, Marc           You Sang To Me  3:50   2000-02-26   
20     2000                 Avant            My First Love  4:28   2000-11-04   
21     2000                 Avant                Separated  4:13   2000-04-29   
22     2000                 BBMak                Back Here  3:35   2000-04-29   
23     2000  Backstreet Boys, The        Shape Of My Heart  3:49   2000-10-14   
24     2000  Backstreet Boys, The  Show Me The Meaning ...  3:54   2000-01-01   
25     2000  Backstreet Boys, The                  The One  3:46   2000-05-27   
26     2000          Badu, Erkyah                 Bag Lady  5:03   2000-08-19   
27     2000              Baha Men     Who Let The Dogs Out  3:17   2000-07-22   
28     2000      Barenaked Ladies                 Pinch Me  3:46   2000-09-09   
29     2000            Beenie Man          Girls Dem Sugar  4:17   2000-10-21   
...     ...                   ...                      ...   ...          ...   
24062  2000          Thomas, Carl                Emotional  4:31   2000-11-25   
24063  2000          Thomas, Carl                   I Wish  3:50   2000-03-25   
24064  2000          Thomas, Carl              Summer Rain  4:57   2000-09-23   
24065  2000         Tippin, Aaron                Kiss This  2:53   2000-08-26   
24066  2000                 Train            Meet Virginia  3:55   1999-10-09   
24067  2000           Trick Daddy                  Shut Up  4:17   2000-05-20   
24068  2000                 Trina                Pull Over  2:58   2000-09-09   
24069  2000         Tritt, Travis       Best Of Intentions  4:15   2000-08-19   
24070  2000               Tuesday                   I Know  4:06   2000-12-30   
24071  2000          Urban, Keith          Your Everything  4:10   2000-07-15   
24072  2000                 Usher            Pop Ya Collar  3:36   2000-11-04   
24073  2000          Vassar, Phil                  Carlene  4:07   2000-03-04   
24074  2000          Vassar, Phil  Just Another Day In ...  3:54   2000-09-30   
24075  2000      Vertical Horizon      Everything You Want  4:01   2000-01-22   
24076  2000      Vertical Horizon             You're A God  3:45   2000-08-26   
24077  2000             Vitamin C  Graduation (Friends ...  4:23   2000-04-15   
24078  2000             Vitamin C                 The Itch  3:30   2000-12-02   
24079  2000          Walker, Clay     

In [7]:
billboard.iloc[0:5, 0:16].groupby(['artist']).mean(1,skipna=True)

UnsupportedFunctionCall: numpy operations are not valid with groupby. Use .groupby(...).mean() instead

In [ ]:
billboard_long.groupby(['year', 'artist', 'track']).mean()

# 직접 해보세요!
## ebola 데이터 집합 살펴보기(155쪽)

In [7]:
ebola = pd.read_csv('../data/country_timeseries.csv')
print(ebola.columns)

Index(['Date', 'Day', 'Cases_Guinea', 'Cases_Liberia', 'Cases_SierraLeone',
       'Cases_Nigeria', 'Cases_Senegal', 'Cases_UnitedStates', 'Cases_Spain',
       'Cases_Mali', 'Deaths_Guinea', 'Deaths_Liberia', 'Deaths_SierraLeone',
       'Deaths_Nigeria', 'Deaths_Senegal', 'Deaths_UnitedStates',
       'Deaths_Spain', 'Deaths_Mali'],
      dtype='object')


In [8]:
print(ebola.iloc[:5, [0, 1, 2, 3, 10, 11]])

         Date  Day  Cases_Guinea  Cases_Liberia  Deaths_Guinea  Deaths_Liberia
0    1/5/2015  289        2776.0            NaN         1786.0             NaN
1    1/4/2015  288        2775.0            NaN         1781.0             NaN
2    1/3/2015  287        2769.0         8166.0         1767.0          3496.0
3    1/2/2015  286           NaN         8157.0            NaN          3496.0
4  12/31/2014  284        2730.0         8115.0         1739.0          3471.0


In [9]:
ebola_long = pd.melt(ebola, id_vars=['Date', 'Day'])
print(ebola_long.head())

         Date  Day      variable   value
0    1/5/2015  289  Cases_Guinea  2776.0
1    1/4/2015  288  Cases_Guinea  2775.0
2    1/3/2015  287  Cases_Guinea  2769.0
3    1/2/2015  286  Cases_Guinea     NaN
4  12/31/2014  284  Cases_Guinea  2730.0


# 직접 해보세요!
## 열 이름 분리하고 데이터프레임에 추가하기(156쪽)

In [10]:
variable_split = ebola_long.variable.str.split('_')

print(variable_split[:5])

0    [Cases, Guinea]
1    [Cases, Guinea]
2    [Cases, Guinea]
3    [Cases, Guinea]
4    [Cases, Guinea]
Name: variable, dtype: object


In [11]:
print(type(variable_split))

<class 'pandas.core.series.Series'>


In [12]:
print(type(variable_split[0]))

<class 'list'>


In [13]:
status_values = variable_split.str.get(0) 
country_values = variable_split.str.get(1)

print(status_values[:5])

0    Cases
1    Cases
2    Cases
3    Cases
4    Cases
Name: variable, dtype: object


In [14]:
print(status_values[-5:])

1947    Deaths
1948    Deaths
1949    Deaths
1950    Deaths
1951    Deaths
Name: variable, dtype: object


In [15]:
print(country_values[:5])

0    Guinea
1    Guinea
2    Guinea
3    Guinea
4    Guinea
Name: variable, dtype: object


In [16]:
print(country_values[-5:])

1947    Mali
1948    Mali
1949    Mali
1950    Mali
1951    Mali
Name: variable, dtype: object


In [17]:
ebola_long['status'] = status_values 
ebola_long['country'] = country_values
print(ebola_long.head())

         Date  Day      variable   value status country
0    1/5/2015  289  Cases_Guinea  2776.0  Cases  Guinea
1    1/4/2015  288  Cases_Guinea  2775.0  Cases  Guinea
2    1/3/2015  287  Cases_Guinea  2769.0  Cases  Guinea
3    1/2/2015  286  Cases_Guinea     NaN  Cases  Guinea
4  12/31/2014  284  Cases_Guinea  2730.0  Cases  Guinea


# 알아두면 좋아요!
## concat 메서드를 응용하여 데이터프레임에 열 추가하기(159쪽)

In [18]:
variable_split = ebola_long.variable.str.split('_', expand=True) 
variable_split.columns = ['status', 'country'] 
ebola_parsed = pd.concat([ebola_long, variable_split], axis=1)

print(ebola_parsed.head())

         Date  Day      variable   value status country status country
0    1/5/2015  289  Cases_Guinea  2776.0  Cases  Guinea  Cases  Guinea
1    1/4/2015  288  Cases_Guinea  2775.0  Cases  Guinea  Cases  Guinea
2    1/3/2015  287  Cases_Guinea  2769.0  Cases  Guinea  Cases  Guinea
3    1/2/2015  286  Cases_Guinea     NaN  Cases  Guinea  Cases  Guinea
4  12/31/2014  284  Cases_Guinea  2730.0  Cases  Guinea  Cases  Guinea


# 직접 해보세요!
## 기상 데이터의 여러 열을 하나로 정리하기 ─ melt, pivot_table 메서드(160쪽)

In [25]:
weather = pd.read_csv('../data/weather.csv') 
print(weather.iloc[:5, :11])

        id  year  month element  d1    d2    d3  d4    d5  d6  d7
0  MX17004  2010      1    tmax NaN   NaN   NaN NaN   NaN NaN NaN
1  MX17004  2010      1    tmin NaN   NaN   NaN NaN   NaN NaN NaN
2  MX17004  2010      2    tmax NaN  27.3  24.1 NaN   NaN NaN NaN
3  MX17004  2010      2    tmin NaN  14.4  14.4 NaN   NaN NaN NaN
4  MX17004  2010      3    tmax NaN   NaN   NaN NaN  32.1 NaN NaN


In [27]:
weather_melt = pd.melt(weather, id_vars=['id', 'year', 'month', 'element'], var_name='day', value_name='temp') 
weather_melt

id  year  month element  day  temp
0    MX17004  2010      1    tmax   d1   NaN
1    MX17004  2010      1    tmin   d1   NaN
2    MX17004  2010      2    tmax   d1   NaN
3    MX17004  2010      2    tmin   d1   NaN
4    MX17004  2010      3    tmax   d1   NaN
5    MX17004  2010      3    tmin   d1   NaN
6    MX17004  2010      4    tmax   d1   NaN
7    MX17004  2010      4    tmin   d1   NaN
8    MX17004  2010      5    tmax   d1   NaN
9    MX17004  2010      5    tmin   d1   NaN
10   MX17004  2010      6    tmax   d1   NaN
11   MX17004  2010      6    tmin   d1   NaN
12   MX17004  2010      7    tmax   d1   NaN
13   MX17004  2010      7    tmin   d1   NaN
14   MX17004  2010      8    tmax   d1   NaN
15   MX17004  2010      8    tmin   d1   NaN
16   MX17004  2010     10    tmax   d1   NaN
17   MX17004  2010     10    tmin   d1   NaN
18   MX17004  2010     11    tmax   d1   NaN
19   MX17004  2010     11    tmin   d1   NaN
20   MX17004  2010     12    tmax   d1  29.9
21   MX17004  2010     12    tmin   d1  13.8
22   MX17004  2010      1    tmax   d2   NaN
23   MX17004  2010      1    tmin   d2   NaN
24   MX17004  2010      2    tmax   d2  27.3
25   MX17004  2010      2    tmin   d2  14.4
26   MX17004  2010      3    tmax   d2   NaN
27   MX17004  2010      3    tmin   d2   NaN
28   MX17004  2010      4    tmax   d2   NaN
29   MX17004  2010      4    tmin   d2   NaN
..       ...   ...    ...     ...  ...   ...
652  MX17004  2010      8    tmax  d30   NaN
653  MX17004  2010      8    tmin  d30   NaN
654  MX17004  2010     10    tmax  d30   NaN
655  MX17004  2010     10    tmin  d30   NaN
656  MX17004  2010     11    tmax  d30   NaN
657  MX17004  2010     11    tmin  d30   NaN
658  MX17004  2010     12    tmax  d30   NaN
659  MX17004  2010     12    tmin  d30   NaN
660  MX17004  2010      1    tmax  d31   NaN
661  MX17004  2010      1    tmin  d31   NaN
662  MX17004  2010      2    tmax  d31   NaN
663  MX17004  2010      2    tmin  d31   NaN
664  MX17004  2010      3    tmax  d31   NaN
665  MX17004  2010      3    tmin  d31   NaN
666  MX17004  2010      4    tmax  d31   NaN
667  MX17004  2010      4    tmin  d31   NaN
668  MX17004  2010      5    tmax  d31   NaN
669  MX17004  2010      5    tmin  d31   NaN
670  MX17004  2010      6    tmax  d31   NaN
671  MX17004  2010      6    tmin  d31   NaN
672  MX17004  2010      7    tmax  d31   NaN
673  MX17004  2010      7    tmin  d31   NaN
674  MX17004  2010      8    tmax  d31  25.4
675  MX17004  2010      8    tmin  d31  15.4
676  MX17004  2010     10    tmax  d31   NaN
677  MX17004  2010     10    tmin  d31   NaN
678  MX17004  2010     11    tmax  d31   NaN
679  MX17004  2010     11    tmin  d31   NaN
680  MX17004  2010     12    tmax  d31   NaN
681  MX17004  2010     12    tmin  d31   NaN

[682 rows x 6 columns]

In [28]:
weather_tidy = weather_melt.pivot_table(
    index=['id', 'year', 'month', 'day'], 
    columns='element', 
    values='temp'
)

print(weather_tidy)

element                 tmax  tmin
id      year month day            
MX17004 2010 1     d1    NaN   NaN
                   d10   NaN   NaN
                   d11   NaN   NaN
                   d12   NaN   NaN
                   d13   NaN   NaN
                   d14   NaN   NaN
                   d15   NaN   NaN
                   d16   NaN   NaN
                   d17   NaN   NaN
                   d18   NaN   NaN
                   d19   NaN   NaN
                   d2    NaN   NaN
                   d20   NaN   NaN
                   d21   NaN   NaN
                   d22   NaN   NaN
                   d23   NaN   NaN
                   d24   NaN   NaN
                   d25   NaN   NaN
                   d26   NaN   NaN
                   d27   NaN   NaN
                   d28   NaN   NaN
                   d29   NaN   NaN
                   d3    NaN   NaN
                   d30  27.8  14.5
                   d31   NaN   NaN
                   d4    NaN   NaN
                   d

In [22]:
weather_tidy_flat = weather_tidy.reset_index() 
print(weather_tidy_flat.head())

element       id  year  month  day  tmax  tmin
0        MX17004  2010      1   d1   NaN   NaN
1        MX17004  2010      1  d10   NaN   NaN
2        MX17004  2010      1  d11   NaN   NaN
3        MX17004  2010      1  d12   NaN   NaN
4        MX17004  2010      1  d13   NaN   NaN


# 직접 해보세요!
## 빌보드 차트의 중복 데이터 처리하기(163쪽)

In [23]:
billboard = pd.read_csv('../data/billboard.csv')
billboard_long = pd.melt(billboard, id_vars=['year', 'artist', 'track', 'time', 'date.entered'], var_name='week', value_name='rating')

print(billboard_long.shape)

print(billboard_long.head())

(24092, 7)
   year        artist                    track  time date.entered week  rating
0  2000         2 Pac  Baby Don't Cry (Keep...  4:22   2000-02-26  wk1    87.0
1  2000       2Ge+her  The Hardest Part Of ...  3:15   2000-09-02  wk1    91.0
2  2000  3 Doors Down               Kryptonite  3:53   2000-04-08  wk1    81.0
3  2000  3 Doors Down                    Loser  4:24   2000-10-21  wk1    76.0
4  2000      504 Boyz            Wobble Wobble  3:35   2000-04-15  wk1    57.0


In [24]:
print(billboard_long[billboard_long.track == 'Loser'].head())

      year        artist  track  time date.entered week  rating
3     2000  3 Doors Down  Loser  4:24   2000-10-21  wk1    76.0
320   2000  3 Doors Down  Loser  4:24   2000-10-21  wk2    76.0
637   2000  3 Doors Down  Loser  4:24   2000-10-21  wk3    72.0
954   2000  3 Doors Down  Loser  4:24   2000-10-21  wk4    69.0
1271  2000  3 Doors Down  Loser  4:24   2000-10-21  wk5    67.0


In [25]:
billboard_songs = billboard_long[['year', 'artist', 'track', 'time']] 
print(billboard_songs.shape)

(24092, 4)


In [26]:
billboard_songs = billboard_songs.drop_duplicates() 
print(billboard_songs.shape)

(317, 4)


In [27]:
billboard_songs['id'] = range(len(billboard_songs)) 
print(billboard_songs.head(n=10))

   year          artist                    track  time  id
0  2000           2 Pac  Baby Don't Cry (Keep...  4:22   0
1  2000         2Ge+her  The Hardest Part Of ...  3:15   1
2  2000    3 Doors Down               Kryptonite  3:53   2
3  2000    3 Doors Down                    Loser  4:24   3
4  2000        504 Boyz            Wobble Wobble  3:35   4
5  2000            98^0  Give Me Just One Nig...  3:24   5
6  2000         A*Teens            Dancing Queen  3:44   6
7  2000         Aaliyah            I Don't Wanna  4:15   7
8  2000         Aaliyah                Try Again  4:03   8
9  2000  Adams, Yolanda            Open My Heart  5:30   9


In [28]:
billboard_ratings = billboard_long.merge( billboard_songs, on=['year', 'artist', 'track', 'time']) 
print(billboard_ratings.shape)

(24092, 8)


In [29]:
print(billboard_ratings.head())

   year artist                    track  time date.entered week  rating  id
0  2000  2 Pac  Baby Don't Cry (Keep...  4:22   2000-02-26  wk1    87.0   0
1  2000  2 Pac  Baby Don't Cry (Keep...  4:22   2000-02-26  wk2    82.0   0
2  2000  2 Pac  Baby Don't Cry (Keep...  4:22   2000-02-26  wk3    72.0   0
3  2000  2 Pac  Baby Don't Cry (Keep...  4:22   2000-02-26  wk4    77.0   0
4  2000  2 Pac  Baby Don't Cry (Keep...  4:22   2000-02-26  wk5    87.0   0


# 직접 해보세요!
## 뉴욕 택시 데이터 준비(166쪽)

In [28]:
import os 
import urllib.request

# 네트워크 상태에 따라 5 ~ 15분이 소요됩니다.
with open('../data/raw_data_urls.txt', 'r') as data_urls:
    for line, url in enumerate(data_urls):
        if line == 5:
            break 
        fn = url.split('/')[-1].strip()
        fp = os.path.join('', '../data', fn)
        print(url)
        print(fp)
        urllib.request.urlretrieve(url, fp)

https://s3.amazonaws.com/nyc-tlc/trip+data/fhv_tripdata_2015-01.csv

../data\fhv_tripdata_2015-01.csv
https://s3.amazonaws.com/nyc-tlc/trip+data/fhv_tripdata_2015-02.csv

../data\fhv_tripdata_2015-02.csv
https://s3.amazonaws.com/nyc-tlc/trip+data/fhv_tripdata_2015-03.csv

../data\fhv_tripdata_2015-03.csv
https://s3.amazonaws.com/nyc-tlc/trip+data/fhv_tripdata_2015-04.csv

../data\fhv_tripdata_2015-04.csv
https://s3.amazonaws.com/nyc-tlc/trip+data/fhv_tripdata_2015-05.csv

../data\fhv_tripdata_2015-05.csv


In [29]:
import glob 
nyc_taxi_data = glob.glob('../data/fhv_*') 
print(nyc_taxi_data)

['../data\\fhv_tripdata_2015-01.csv', '../data\\fhv_tripdata_2015-02.csv', '../data\\fhv_tripdata_2015-03.csv', '../data\\fhv_tripdata_2015-04.csv', '../data\\fhv_tripdata_2015-05.csv']


In [30]:
taxi1 = pd.read_csv(nyc_taxi_data[0]) 
taxi2 = pd.read_csv(nyc_taxi_data[1]) 
taxi3 = pd.read_csv(nyc_taxi_data[2]) 
taxi4 = pd.read_csv(nyc_taxi_data[3]) 
taxi5 = pd.read_csv(nyc_taxi_data[4])

In [31]:
print(taxi1.head(n=2)) 
print(taxi2.head(n=2)) 
print(taxi3.head(n=2)) 
print(taxi4.head(n=2)) 
print(taxi5.head(n=2))

  Dispatching_base_num          Pickup_date  locationID
0               B00013  2015-01-01 00:30:00         NaN
1               B00013  2015-01-01 01:22:00         NaN
  Dispatching_base_num          Pickup_date  locationID
0               B00013  2015-02-01 00:00:00         NaN
1               B00013  2015-02-01 00:01:00         NaN
  Dispatching_base_num          Pickup_date  locationID
0               B00029  2015-03-01 00:02:00       213.0
1               B00029  2015-03-01 00:03:00        51.0
  Dispatching_base_num          Pickup_date  locationID
0               B00001  2015-04-01 04:30:00         NaN
1               B00001  2015-04-01 06:00:00         NaN
  Dispatching_base_num          Pickup_date  locationID
0               B00001  2015-05-01 04:30:00         NaN
1               B00001  2015-05-01 05:00:00         NaN


In [32]:
print(taxi1.shape) 
print(taxi2.shape) 
print(taxi3.shape) 
print(taxi4.shape) 
print(taxi5.shape)

(2746033, 3)
(3126401, 3)
(3281427, 3)
(3917789, 3)
(4296067, 3)


In [33]:
taxi = pd.concat([taxi1, taxi2, taxi3, taxi4, taxi5])

print(taxi.shape)

(17367717, 3)


In [45]:
taxi['Pickup_date']

0          2015-01-01 00:30:00
1          2015-01-01 01:22:00
2          2015-01-01 01:23:00
3          2015-01-01 01:44:00
4          2015-01-01 02:00:00
5          2015-01-01 02:00:00
6          2015-01-01 02:00:00
7          2015-01-01 02:50:00
8          2015-01-01 04:45:00
9          2015-01-01 06:30:00
10         2015-01-01 07:15:00
11         2015-01-01 07:54:00
12         2015-01-01 10:30:00
13         2015-01-01 11:00:00
14         2015-01-01 12:42:00
15         2015-01-01 12:58:00
16         2015-01-01 13:28:00
17         2015-01-01 14:27:00
18         2015-01-01 16:15:00
19         2015-01-01 16:59:00
20         2015-01-01 17:53:00
21         2015-01-01 18:45:00
22         2015-01-01 19:00:00
23         2015-01-01 19:54:00
24         2015-01-01 21:45:00
25         2015-01-01 22:00:00
26         2015-01-02 01:00:00
27         2015-01-02 01:08:00
28         2015-01-02 01:10:00
29         2015-01-02 02:45:00
                  ...         
4296037    2015-05-31 08:15:00
4296038 

In [42]:
taxi['Pickup_date'].iloc[-8:] >= '00:00:00' and taxi['Pickup_date'].iloc[0][-8:] <= '05:00:00']

KeyError: True

'00:30:00'

# 알아두면 좋아요!
## 반복문으로 데이터 준비하기(169쪽)

In [39]:
list_taxi_df = [] 

for csv_filename in nyc_taxi_data:
    # print(csv_filename)
    df = pd.read_csv(csv_filename)
    list_taxi_df.append(df) 

print(len(list_taxi_df))

5


In [40]:
print(type(list_taxi_df[0]))

<class 'pandas.core.frame.DataFrame'>


In [41]:
print(list_taxi_df[0].head())

  Dispatching_base_num          Pickup_date  locationID
0               B00013  2015-01-01 00:30:00         NaN
1               B00013  2015-01-01 01:22:00         NaN
2               B00013  2015-01-01 01:23:00         NaN
3               B00013  2015-01-01 01:44:00         NaN
4               B00013  2015-01-01 02:00:00         NaN


In [42]:
taxi_loop_concat = pd.concat(list_taxi_df) 
print(taxi_loop_concat.shape)

(17367717, 3)


In [43]:
print(taxi.equals(taxi_loop_concat))

True
